kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/BootstrapServersAnnotation.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kubernetes.operator;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.api.model.ObjectMetaFluent;
+
+import io.kroxylicious.proxy.tag.VisibleForTesting;
+
+/**
+ * Used to annotate kubernetes Services with the client facing bootstrap servers that we expect clients will
+ * connect with. We include (clusterName, ingressName, bootstrapServers) for each bootstrap address that the
+ * client should eventually host.
+ * <p>
+ * We also include a version in the value of the annotation, in case we ever need to break compatibility but
+ * continue reading older Service's metadata.
+ * </p>
+ */
+public class BootstrapServersAnnotation {
+
+    public static final String BOOTSTRAP_SERVERS_ANNOTATION_KEY = "kroxylicious.io/bootstrap-servers";
+
+    private BootstrapServersAnnotation() {
+    }
+
+    /**
+     * Adds a `kroxylicious.io/bootstrap-servers`annotation to the supplied metadata fluent
+     * @param meta the Metadata fluent builder to add the annotation to
+     * @param bootstrapServers the bootstrap servers to serialize into the annotation value
+     */
+    public static void annotate(ObjectMetaFluent<?> meta, Set<BootstrapServer> bootstrapServers) {
+        if (bootstrapServers.isEmpty()) {
+            return;
+        }
+        meta.addToAnnotations(BOOTSTRAP_SERVERS_ANNOTATION_KEY, toAnnotation(bootstrapServers));
+    }
+
+    /**
+     * Read bootstrap servers from ObjectMeta, extracting them from an `kroxylicious.io/bootstrap-servers`
+     * annotation if present.
+     * @param meta the metadata to extract the bootstrap servers from
+     * @return the bootstrap servers from the metadata if the annotation is present, else an empty Set
+     */
+    public static Set<BootstrapServer> bootstrapServersFrom(ObjectMeta meta) {
+        if (!meta.getAnnotations().containsKey(BOOTSTRAP_SERVERS_ANNOTATION_KEY)) {
+            return Set.of();
+        }
+        else {
+            return fromAnnotation(meta.getAnnotations().get(BOOTSTRAP_SERVERS_ANNOTATION_KEY));
+        }
+    }
+
+    @JsonPropertyOrder({ "clusterName", "ingressName", "bootstrapServers" })
+    public record BootstrapServer(String clusterName, String ingressName, String bootstrapServers) {
+        public BootstrapServer {
+            Objects.requireNonNull(clusterName);
+            Objects.requireNonNull(ingressName);
+            Objects.requireNonNull(bootstrapServers);
+        }
```

Comment:
Feels a little recursive for a `BootstrapServer` to contain a `BootstrapServers`, note the plural. Maybe the string should be `upstreamAddresses` or the record should be `BootstrapAddressComponents` ?

Reply:
The plural is more trying to keep it aligned with kafka `bootstrap.servers` which we've decided to do elsewhere, in the KafkaService CRD, though in the ingress status we use `bootstrapServer`. Maybe we can go with singular, since we're using it in the ingress status.

I think it's the record name that needs work, maybe you could say that these are client-facing bootstrap servers "hosted" by the Service. So it might be more like `HostedBootstrapServers`? I've used client-facing a few times in code,bit long but I think quite clear that it's for clients to consume. `ClientFacingBootstrapServer`?

Reply:
Yeah I think the consistency point is good and the record name is the one that needs work.

I generally aim for the shortest explicit name